File: buckets.py
import numpy as np
import math as math
import random
from datetime import datetime

def bucket(I, y):
    #calculating the probabilities for each point in the distribution
    s = 0
    for val in I:
        s = s+val
        
    probabilities = list()
    for val in I:
        probabilities.append((val/s)*100)
    
    #making probability buckets
    buckets = list()            #master list of all buckets of same probability
    bucket_intensity = list()   #corresponding list of intensities for each bucket
    current_bucket = list()
    current_intensity = list()
    current_sum = 0             #to check for probability
    
    x = 0
    while x < len(probabilities):
        prob = probabilities[x]
        current_bucket.append(y[x])
        current_intensity.append(I[x])
        current_sum = current_sum + prob
        
        #if probability is pretty close to 1, make this a new bucket
        if (1-current_sum < 0.01):
            buckets.append(current_bucket)
            bucket_intensity.append(current_intensity)
            current_sum = 0
            current_bucket = list()
            current_intensity = list()
        x = x+1
    
    #selecting a bucket randomly
    #as each should have the same probability
    random.seed(datetime.now())
    num = random.randint(0,len(buckets))
    if(num == len(buckets)):
        num = 0
    bucket = buckets[num]
    intensity = bucket_intensity[num]

    return [bucket, intensity]

def call_bucket(values, x_vals):
   bucket1 = bucket(values, x_vals)
   intensity1 = bucket_info[0]     #these are named 1 so intensity and x_vals don't get overwritten
   x_vals1 = bucket_info[1]
   while( len(x_vals1) > 1 ):      #break buckets into buckets until only 1 thing in it
      bucket_info = B.bucket(intensity1, x_vals1)
      intensity1 = bucket_info[0]
      x_vals1 = bucket_info[1]
   if random.randint(0,10)%2 == 0: #because for some reason otherwise they're only negative
      return x_vals1[0]
   else:
      return -1*x_vals1[0]

File: buckets.cpp
#include "buckets.h"
#include <iostream>
#include <list>
#include <stdlib.h>
#include <time.h>
using namespace std;

list<list<float> > bucket(list<float> intensity, list<float> y)
{
	list<list<float> > result;

	// calculating the probabilities for each point in the distribution
	float sum = 0;
	for(list<float>::iterator itr = intensity.begin(); itr != intensity.end(); itr++)
		sum += *itr;
	
	list<float> probabilities;
	for(list<float>::iterator itr = intensity.begin(); itr != intensity.end(); itr++)
		probabilities.push_back((*itr/sum)*100);

	// making the probability buckets
	list<list<float> > buckets;
	list<list<float> > bucket_intensity;
	list<float> current_bucket;
	list<float> current_intensity;
	float current_sum = 0;

	list<float>::iterator itr1 = intensity.begin();
	list<float>::iterator itr2 = probabilities.begin();
	list<float>::iterator itr3 = y.begin();

	for(; itr1 != intensity.end(); itr1++, itr2++, itr3++)
	{
		float prob = *itr2;
		current_bucket.push_back(*itr3);
		current_intensity.push_back(*itr1);
		current_sum += prob;
		if(1-current_sum < 0.01)
		{
			buckets.push_back(current_bucket);
			bucket_intensity.push_back(current_intensity);
			current_sum = 0;
			current_bucket.clear();
			current_intensity.clear();
		}
	}

	// selecting a particular bucket at random
	// as they should have the same probability
	// of finding the particle there
	srand(time(NULL));
	int num = rand() % buckets.size();
	list<list<float> >::iterator ij = buckets.begin();
	list<list<float> >::iterator ji = bucket_intensity.begin();
	for(int i = 0; i < num; i++)
	{
		ij++;
		ji++;
	}
	result.push_back(*ij);
	result.push_back(*ji);

	return result;
}